### ISSUE TITLE
[clang-tidy] False positive from misc-redundant-expression with dependent conditionals

### ISSUE THREAD

**Issue by al42and** [llvm/llvm-project#57726]
In the following code, `ALLOW_ARGS` is a PP macro. When it is set to 0, clang-tidy complains about the redundant expression (annotated).

Technically, the expression is indeed redundant in this case, but the parent condition makes it clear that we are not interested in this comparison when `sc_allowArgs` is false.

```cpp
static constexpr bool sc_allowArgs = (ALLOW_ARGS != 0);
static constexpr int sc_maxArgs = sc_allowArgs ? 10 : 0;

int main(int argc, char **argv) {
  if constexpr (sc_allowArgs) {
    if (argc < 0 || argc >= sc_maxArgs) { // <<
      return -1;
    }
  }
  return 0;
}
```

```
$ clang-tidy-15 -checks=misc-redundant-expression test.cpp -- -DALLOW_ARGS=0 -std=c++17
1 warning generated.
/tmp/test.cpp:6:18: warning: logical expression is always true [misc-redundant-expression]
    if (argc < 0 || argc >= sc_maxArgs) {
```

Godbolt: https://godbolt.org/z/3P9914MWb

**Comment by JonasToth** [comment]
not sure if that can be properly generalized to be true always and not just in your specific example.

**Comment by njames93** [comment]
I suppose the only thing we can really do here to address this is not traverse into false if constexpr branches.
